odoo/tools/lru.py

```diff
@@ -15,48 +14,99 @@ class LRU(MutableMapping[K, V], typing.Generic[K, V]):
     """
     Implementation of a length-limited O(1) LRU map.
 
-    Original Copyright 2003 Josiah Carlson, later rebuilt on OrderedDict and added typing.
+    We use ``dict`` for entries and another one to keep ordering of keys.
     """
+
+    __slots__ = ('_count', '_lock', '_ordering', '_values')
+
     def __init__(self, count: int, pairs: Iterable[tuple[K, V]] = ()):
+        assert count > 0, "LRU needs a positive count"
+        self._count = count
         self._lock = threading.RLock()
-        self.count = max(count, 1)
-        self.d: collections.OrderedDict[K, V] = collections.OrderedDict()
+        self._values: dict[K, V] = {}
+        #
+        # The dict self._values contains the LRU items, while self._ordering
+        # only keeps track of their order, the most recently accessed ones
+        # appearing last. For performance reasons, we only use the lock when
+        # modifying the LRU, while reading it is lock-free (and thus faster).
+        #
+        # This strategy may result in inconsistent order. Inside self._ordering,
+        # concurrently accessed keys may be missing, but will eventually be
+        # added. This could result in keys being added back after their removal
+        # from the LRU. This results in the following invariant:
+        #
+        #     self._values <= self._ordering | "keys being accessed"
+        #
+        self._ordering: dict[K, None] = {}
+
+        # Initialize
         for key, value in pairs:
             self[key] = value
 
-    @locked
-    def __contains__(self, obj: K) -> bool:
-        return obj in self.d
+    @property
+    def count(self) -> int:
+        return self._count
+
+    def __contains__(self, key: object) -> bool:
+        return key in self._values
 
-    @locked
-    def __getitem__(self, obj: K) -> V:
-        a = self.d[obj]
-        self.d.move_to_end(obj, last=False)
-        return a
+    def __getitem__(self, key: K) -> V:
+        val = self._values[key]
+        # move key at the last position in self._ordering
+        self._ordering[key] = self._ordering.pop(key, None)
+        return val
 
-    @locked
-    def __setitem__(self, obj: K, val: V):
-        self.d[obj] = val
-        self.d.move_to_end(obj, last=False)
-        while len(self.d) > self.count:
-            self.d.popitem(last=True)
+    def __setitem__(self, key: K, val: V):
+        values = self._values
+        ordering = self._ordering
+        with self._lock:
+            values[key] = val
+            ordering[key] = ordering.pop(key, None)
+            while True:
+                # if we have too many keys in ordering, filter them out
+                if len(ordering) > len(values):
```

Comment:
could you explain when `len(ordering) > len(values)`?
I think `len(ordering) <= len(values)` 🤔 And even if `len(ordering) > len(values)` we still could use the code below to handle the `count` since `values.pop(key, None)` a non-existing key won't change its `len`

Reply:
`ordering` contains `values` except those that are being accessed concurrently that are added back after the access. If you pop or delete a value while concurrently accessing it, you will add that value to `ordering` and it would disappear from `values`. This detects such a case in a fast-way before it becomes an issue (ordering growing indefinitely) and will compact `ordering`.

Reply:
It's this situation:
```
thread A                                    thread B

__getitem__(key):                           ⋮
    val = self._values[key]                 ⋮
    z = self._ordering.pop(key, None)       ⋮
    ⋮                                       pop(key)
    self._ordering[key] = z                 ⋮
```
After that, `key` is missing from `_values` but is present in `_ordering`.  This may eventually lead to `_ordering` containing more keys than `_values`.